exercise/static/exercise/chapter.js

```diff
@@ -234,40 +357,207 @@
 					});
 				}
 			}
+
 			$.augmentExerciseGroup(content);
 			window.postMessage({
 				type: "a-plus-bind-exercise",
 				id: this.chapterID
 			}, "*");
 		},
 
-		submit: function(form_element) {
-			//$(form_element).find(":input").prop("disabled", true);
-			//this.showLoader("submit");
-			this.chapter.openModal(this.chapter.messages.submit);
+		// Submit the formData to given url and then execute the callback.
+		submitAjax: function(url, formData, callback, retry) {
 			var exercise = this;
-			$.ajax($(form_element).attr("action"), {
+			$.ajax(url, {
 				type: "POST",
-				data: new FormData(form_element),
+				data: formData,
 				contentType: false,
 				processData: false,
 				dataType: "html"
-			}).fail(function() {
+			}).fail(function(xhr, textStatus, errorThrown) {
+					// Retry a few times if submission is not successful
+					retry = retry || 0;
+					if (xhr.status !== 200 && retry < 5) {
+						setTimeout(
+							function() {
+								exercise.submitAjax(url, formData, callback, retry + 1);
+							}, 100);
+					}
 				//$(form_element).find(":input").prop("disabled", false);
 				//exercise.showLoader("error");
-				this.chapter.modalError(exercise.chapter.messages.error);
-			}).done(function(data) {
-				//$(form_element).find(":input").prop("disabled", false);
-				//exercise.hideLoader();
-				var input = $(data);
-				if (exercise.quiz) {
-					var badge = input.find('.badge').eq(2).clone();
-					exercise.update(input);
-					exercise.chapter.modalSuccess(exercise.element, badge);
+
+				if (!exercise.active_element) {
+					exercise.chapter.modalError(exercise.chapter.messages.error);
+				} else {
+					// active elements don't use loadbar so the error message must be shown
+					// in the element container
+					var feedback = $("<div>");
+					feedback.attr('id', 'feedback');			
+					feedback.append(exercise.chapter.messages.error);
+					exercise.updateOutput(feedback);
+				}
+			}).done(function (data) {
+				/* Bad fix for database locked problem when saving submitted files:
+					 test if the data contains message that the submission was not saved and resubmit
+				*/
+				retry = retry || 0;
+				if ($(data).find("div:contains('The submission could not be saved for some reason')").length > 0 && retry < 5) {
+					console.log("Submission not saved: trying submitAjax again in 100ms");
```

Comment:
When one input triggers and submits several active elements, the multiple submits sometimes caused a "database locked" error in the SQLite database, and if there are any file submit exercises that fail, the submitted files are not stored because of the error. (Apparently the same database locked error returns an error code if the submission does not include files added by  the user.)

Reply:
Does this happen only with SQLite and not with real DB? Are you running multiple django processes with SQLite?

Reply:
Yes it does seem to be a problem with SQLite only, though it hasn't been tested that extensively. The setup has been pretty much out-of-the-box A+ & mooc-grader with the addition of active element JS that can send multiple AJAX requests to the exercise urls (still single request per exercise) when the user clicks submit.

Reply:
Apparently django runserver use threads, so when running with sqlite it can try to open the file at the same time. Thus, we need to disable threads with sqlite. That can be done with runserver flag `--nothreading`. That at least should help with the case.

If you could run your test environment with that flag on (or `command: "python3 manage.py runserver 0.0.0.0:8000 --nothreading"` in docker-compose.yml) and inform if this error still appears (or doesn't seem to). If that flag fixes it, then I think it could be turned on by default in our test containers (run-aplus-front).

Any case, you can remove this part of the JS as it's not required in production.

Reply:
Everything seems to work fine with the flag, thanks!